hbase-server/src/main/java/org/apache/hadoop/hbase/master/HMaster.java

```diff
@@ -4088,6 +4090,12 @@ public void onConfigurationChange(Configuration newConf) {
     } catch (IOException e) {
       LOG.warn("Failed to initialize SuperUsers on reloading of the configuration");
     }
+    // update region server coprocessor if the configuration has changed.
+    if (CoprocessorConfigurationUtil.checkConfigurationChange(getConfiguration(), newConf,
+      CoprocessorHost.MASTER_COPROCESSOR_CONF_KEY)) {
+      LOG.info("Update the master coprocessor(s) because the configuration has changed");
+      this.cpHost = new MasterCoprocessorHost(this, newConf);
+    }
```

Comment:
Don't we need to handle maintenance mode here?

And how are we ensuring that newConf includes the System coprocessor we are adding based on whether the feature is on or not?
For eg: QuotasObserver
conf.setStrings(CoprocessorHost.MASTER_COPROCESSOR_CONF_KEY, MasterQuotasObserver.class.getName());

Another is RSGroupAdminEndpoint?

Reply:
I missed that, but I found that `maintenanceMode` should not be in the picture because `cpHost` only initialized when  `maintenanceMode=false`

your second part of question is a harder one, in other words, you are asking if there is any existing CP that uses a boolean flag instead of using coprocessor class configuration in hbase-site. if so, how do we handle such override. 

I don't have a good idea but at least for the `MasterQuotasObserver`, I can handle it by adding block of  `updateConfigurationForQuotasObserver` before the reload of the master coprocessor host. 

`RSGroupAdminEndpoint` should be deprecated, and it's a new object `rsGroupInfoManager` if the coprocessor class name was found as part of the `MASTER_COPROCESSOR_CONF_KEY` . we can also handle it as part of the reload. 

but for a more general CP that has the boolean switch, this change may break them. 

Would you suggest we don't support master coprocessor with dynamic configuration ?

Reply:
after a second thought on the `rsGroupInfoManager`, it should not be reloaded when the configuration change because it's not a coprocessor but instead a core executor within the HMaster. 

if the user really want to disable it, the user will need to turn off by setting `hbase.balancer.rsgroup.enabled=false` and remove `RSGroupAdminEndpoint` from the `MASTER_COPROCESSOR_CONF_KEY` , then perform a hard restart. what do you think?